pkg/ssh/connection_golang.go

```diff
@@ -272,23 +273,54 @@ func ValidateAndConfigure(uri *url.URL, iden string) (*ssh.ClientConfig, error)
 	if err != nil {
 		return nil, err
 	}
-	keyFilePath := filepath.Join(homedir.Get(), ".ssh", "known_hosts")
-	known, err := knownhosts.New(keyFilePath)
-	if err != nil {
-		return nil, fmt.Errorf("creating host key callback function for %s: %w", keyFilePath, err)
-	}
 
+	var keyErr *knownhosts.KeyError
 	cfg := &ssh.ClientConfig{
-		User:            uri.User.Username(),
-		Auth:            authMethods,
-		HostKeyCallback: known,
-		Timeout:         tick,
+		User: uri.User.Username(),
+		Auth: authMethods,
+		HostKeyCallback: ssh.HostKeyCallback(func(host string, remote net.Addr, pubKey ssh.PublicKey) error {
+			keyFilePath := filepath.Join(homedir.Get(), ".ssh", "known_hosts")
+			known, err := knownhosts.New(keyFilePath)
+			if err != nil {
+				if os.IsNotExist(err) && insecureIsMachineConnection {
+					logrus.Warn("please create a known_hosts file and recreate the machine to add the public key entry")
+					return nil
+				}
+				return err
+			}
+			// we need to check if there is an error from reading known hosts for this public key and if there is an error, what is it, and why is it happening?
+			// if it is a key mismatch we want to error since we know the host using another key
+			// however, if it is a general error not because of a known key, we want to add our key to the known_hosts file
+			hErr := known(host, remote, pubKey)
+			// if keyErr.Want is not empty, we are receiving a different key meaning the host is known but we are using the wrong key
+			switch {
+			case errors.As(hErr, &keyErr) && len(keyErr.Want) > 0:
+				logrus.Warnf("ssh host key mismatch for host %s, got key %s of type %s", host, ssh.FingerprintSHA256(pubKey), pubKey.Type())
+				return keyErr
+			// if keyErr.Want is empty that just means we do not know this host yet, add it.
+			case errors.As(hErr, &keyErr) && len(keyErr.Want) == 0:
+				// write to known_hosts
+				knownFile, err := writeKnownHosts(host, pubKey)
+				if err != nil {
+					if os.IsNotExist(err) {
+						logrus.Warn("podman will soon require a known_hosts file to function properly.")
```

Comment:
If this code is confident enough to in the public key to add an entry, why would it be reluctant to create a new file with a single line?

Reply:
I thought our desired path was to warn people for a few podman versions and then maybe talk about creating a file? I think this is exiting the scope of this PR.

This fix is needed for 4.3 next week

Reply:
I agree with @mtrmac. The stock behavior for podman machine is that a file does not need to be created. Granted, it was because it wasn't in use, but from a UX perspective, we should avoid having podman machine instances break after the upgrade. This is also the stock behavior for stock ssh (silent file creation). Change-wise should be safe and small (just adding O_CREATE to flags)

Reply:
@n1hility This path is not really a `podman machine` concern in _this_ sense. If the file is _entirely missing_, we get into the other path with the `insecureIsMachineConnection` check.

Even this path does not _break_ machines, it warns and continues.

It’s the non-machine case where the behavior between existing file with no match / no file is hard for me to understand.

Reply:
Actually, my original comment is way off base.

If we get to the path that start with a `known()` call at all, we _know_ that the file did exist when creating `known`. So the only way we can get here is if someone removed `known_hosts`, or a parent directory, between the read and this write.

In that case I don’t think we _really_ need any special handling for `os.IsNotExist` (though using `O_CREATE` _would_, I think, be closer to the right thing to do [with the absolutely right thing being worrying about locking the `known_hosts` writers across disparate implementations??? ugh]).

But also, the current warning text seems very unrelated to the situation at hand: it is warning about a `known_hosts` file that did exist just a few milliseconds ago.

Reply:
> I thought our desired path was to warn people for a few podman versions and then maybe talk about creating a file? I think this is exiting the scope of this PR

AFAICS, this is not what the current implementation (without `insecureIsMachineConnection` set) does: there is not a warning, but a hard failure. From a security point of view, I don’t care, the hard failure is secure. The UX might be bad.

Or are you talking about “our desired path” for `podman machine`? Nothing in _this_ code path checks for `insecureIsMachineConnection`, so I don’t see how that relates.

Is there a record of that desired path I should be following?

---

Obviously _something_ about this approach to writing the code is not working.

What’s some practical and proven way to capture the design intent, so that it can then be verified against code? (No, not formal proof languages. Something much higher-level, maybe at the level of sample CLI sessions, hopefully less detailed than that.) 

Failing any better ideas, if the intent is to warn about $foo in $situation, and then replace that warning later, can that be captured in #1152? I don’t think that checklist is the right way to discuss this but it’s better than nothing at all.

Reply:
@mtrmac @cdoern  aha sorry! I just skimmed the shown snippets without looking at the full code. Ok, that alleviates my primary concern. 

That said I agree  it would make sense to auto-create in the non-machine case as well, and this could simplify the logic a bit. 

>_ need any special handling for `os.IsNotExist` (though using `O_CREATE` _would_, I think, be closer to the right thing to do [with the absolutely right thing being worrying about locking the `known_hosts` writers across disparate implementations??? ugh]).

I think its ok relying on O_APPENDs atomicity as long as its always one write for the record. I just double checked the openssh source and they dont use cooperative file locking on known hosts, so its effect would be limited to just concurrent usage of podman

Reply:
I am mocking up a version where we create a known_hosts file which removes the need for `insecureIsMachineConnection`

Reply:
> I think its ok relying on O_APPENDs atomicity as long as its always one write for the record. I just double checked the openssh source and they dont use cooperative file locking on known hosts

Thanks for taking the time to actually check. Yes, looking at `add_host_to_hostfile` in https://cvsweb.openbsd.org/cgi-bin/cvsweb/src/usr.bin/ssh/hostfile.c?rev=1.93&content-type=text/x-cvsweb-markup , we don’t need to do any better than that:
- Just write text+`\n`, don’t worry about files that don’t end with a newline.
- Use O_CREATE | O_APPEND, don’t do anything special about concurrency (but potentially create `~/.ssh`, and in that case it is _fairly important_ to use mode 0700).

I guess that’s for a future version?

Reply:
@mtrmac do you think we should merge this as is for 4.3? I modified it locally to create ~/.ssh/known_hosts, I can push that if needed.

Reply:
@cdoern I apologize, for some reason I thought that our 4.2 baseline is https://github.com/containers/common/pull/1094 , and that this PR is a net decrease in security. I have only today actually checked, and 4.2 has the pre-c/common ssh implementation if I’m not mistaken.

Assuming that’s the case, any version of this is an improvement over 4.2. So, at this point I think the priorities are, in this order:
- Make sure this + the corresponding Podman PRs is merged into 4.3, so that we have a working Podman machine in 4.3.
- Get the implementation of the feature set _as is right now_ correct and usable. (I think the PR as is is fine _in isolation_, but can we rely on tests for all relevant features? I.e. I think that rather than add more last-minute features, it would be more useful to do a manual walk-through of all the relevant use cases [and to capture that for posterity], and ensure that we haven’t overlooked any corner cases / interactions. (E.g. is there really no intersection between `ssh_native` and the `IsMachineConnection` cases?)

And for 4.4, let’s first document what we want to do, then see what pieces are missing, and finally finish this.